Add resource option to configure unused pins as pulldown

schugabe opened the pull request:
All pins are initialized to pullup inputs. With `resource PULLDOWN 1 PIN` it is possible to reconfigure the pin so it is a pulldown input.

With this it's possible to prevent errors on some boards that have multiple pins connected
with an inverter and the pullup flows back via the inverter and pulls other pins low (e.g. #7849)

I'm not certain if the `resource PULLUP` is necessary but added it to offer both options.

Currently each resource can be assigned to two differnt pins. Do you think this should be a differnt number?

mikeller commented on src/main/pg/pin_up_down.h:
```diff
@@ -0,0 +1,33 @@
+    ioTag_t ioTag[PIN_UP_DOWN_COUNT];
+} pinUpDownConfig_t;
+
+PG_DECLARE(pinUpDownConfig_t, pinPullupConfig);
```
Instead of declaring a parameter group that consists of an array of elements, it is better to declare a parameter group array of the elements. See `pg/timerio.[ch]` for an example of how this can be done.

mikeller commented on src/main/target/common_pre.h:
```diff
@@ -264,6 +264,7 @@
 #define USE_HUFFMAN
 #define USE_PINIO
 #define USE_PINIOBOX
+#define USE_PIN_UP_DOWN
```
Instead of adding this here, it is better to add this to the `#if (FLASH_SIZE > 256)` below, as I don't think there are any F3 boards that need this, and flash space on F3 is very limited.

mikeller commented:
I think we can go for 4 pins of each type we do - no point in skimping, and then have users run out of pins - this functionality will also be useful for things like 'pinning' bright status LEDs on some integrated micro boards.

Also, while I do not know of any application that needs a (strong-ish) pull up, I think having one available will be useful.

@jflyper: What's your thoughts?

ledvinap commented:
IMO this is overkill. The function can be implemented using PINIO, just don't connect pin to function.
Simply port new pullup/pulldown functionality to it ...

schugabe commented:
I'd say PINIO is overkill for what it is often needed. This is a much simpler syntax.

I added "pull" to the names to make the function more clear, reordered the includes and changed the pg config to an array type and fixed the places with the wrong defines. I hope I did not miss any suggested changes.

mikeller commented on src/main/pg/pin_pull_up_down.c:
```diff
@@ -0,0 +1,42 @@
+
+void pgResetFn_pinPullupConfig(pinPullUpDownConfig_t *config)
+{
+    config->ioTag = IO_TAG(NONE);
```
Since you switched this to an array parameter group, the expectation is that the reset function resets all elements of the array, see `pg/timerio.c`.

Also, you can probably declare a common function, and call it from both reset functions.

schugabe replied:
I assumed (without checking) that the reset function is called for each element. I added the reset for all elements now.

hydra commented:
adding pull/up down functionality to existing PINIO seems better to me.  We already know we need UI for PINIO, with this we'd also need new UI for this too.  As I user I might also be confused as to which system, this or pinio, I should be using.

@schugabe did you consider using PINIO functionality for this, if so what reasons did you have for going in this direction?

hydra commented:
@ledvinap and @schugabe can you both give details of why you think each approach you don't like is 'overkill'.  There's quite a bit of disagreement on this PR, not just here but in Slack too, so getting some extra thoughts on this might help clarify things.

mikeller commented:
@hydra:

> adding pull/up down functionality to existing PINIO seems better to me. We already know we need UI for PINIO, with this we'd also need new UI for this too.

Adding pinio to the UI is exactly what concerns me - this is intended to be used to make certain boards with 'exotic' solutions for inverters and other on-board peripherals work. As such, exposing it to users in a way that makes it non-obvious what it is meant for will only introduce the possibility for users to inadvertently change it when they are setting up pinio, and then ending up with a 'broken' board.
I guess the same problem will have to be solved when we are adding the rest of the resource configuration to the UI - we will need a way to have a 'no user serviceable parts inside' part of the configuration that stops users from changing settings that will 'break' their board - at the moment the CLI does a good job of creating a barrier for this.

schugabe commented:
@hydra I did not see anything about this PR on slack? 

My main problem with pinio is that the name of the resource is vague. I was searching for a function to configure a pin as pulldown but reading resource pinio did not sound like something that could be used for this task. And after reading the wiki page you have to think about setting the correct bits in pinio_config and if you scroll over an cli dump it is also not obvious what the config does.

One way to solve this would be to add another cli command to configure pinio more explicit. But that seems kind of overkill for this feature.

If you simply consider a pulldown as a resource and connect the pin to it. You need some additional config space but other than that I would call it more elegant and in line with the cli syntax we are already using.
